model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,668 @@
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import static org.keycloak.models.UserSessionModel.CORRESPONDING_SESSION_ID;
+import static org.keycloak.models.UserSessionModel.SessionPersistenceState.TRANSIENT;
+import static org.keycloak.models.map.userSession.SessionExpiration.setClientSessionExpiration;
+import static org.keycloak.models.map.userSession.SessionExpiration.setUserSessionExpiration;
+import static org.keycloak.utils.StreamsUtil.paginatedStream;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public class MapUserSessionProvider implements UserSessionProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapUserSessionProvider.class);
+    private final KeycloakSession session;
+    protected final MapKeycloakTransaction<UUID, MapUserSessionEntity, UserSessionModel> userSessionTx;
+    protected final MapKeycloakTransaction<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionTx;
+    private final MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore;
+    private final MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore;
+
+    /**
+     * Storage for transient user sessions which lifespan is limited to one request.
+     */
+    private final Map<UUID, MapUserSessionEntity> transientUserSessions = new HashMap<>();
+
+    public MapUserSessionProvider(KeycloakSession session, MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore,
+                                  MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore) {
+        this.session = session;
+        this.userSessionStore = userSessionStore;
+        this.clientSessionStore = clientSessionStore;
+        userSessionTx = userSessionStore.createTransaction(session);
+        clientSessionTx = clientSessionStore.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(userSessionTx);
+        session.getTransactionManager().enlistAfterCompletion(clientSessionTx);
+    }
+
+    private Function<MapUserSessionEntity, UserSessionModel> userEntityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return (origEntity) -> {
+            if (origEntity.getExpiration() <= Time.currentTime()) {
```

Comment:
I see what you mean. But I wouldn't be much worried because the only place where this method is called from is `UserSessionManager.createOrUpdateOfflineSession` An offline user session is obtained and if it is not found or it is expired, the new one is created. Then immediately the `createOfflineClientSession` is called if there is no offline client session present.  
Also the `removeExpired` is never called from `services` so it doesn't affect this method. 

Regarding your first question I think it could be possible to implement this functionality in `getUpdatedNotRemoved / read` methods but I'm not sure how much effort it would require. In any case I would prefer not to deal with that in this PR unless it's a tiny change.

Reply:
This was just an example to explain what I meant. Does it help that `removeExpired` is not called from `services`? The issue is that previously an expired session was never loaded from the store because it wasn't present (is it correct?) and now, we can do some code that wouldn't be executed before.

I created https://issues.redhat.com/browse/KEYCLOAK-17884, feel free to add more details or remove if I misunderstood something.